### PR TITLE
Add Locale Extensions retain_by_type

### PR DESCRIPTION
**Pull request by sffc**
Progress on #1109 
Progress on #1703

It may be useful to review this in tandem with #1737.

### CODE REVIEW

File: components/locid/src/extensions/mod.rs [hunk 1]
```diff
@@ -135,6 +135,45 @@ impl Extensions {
...
+    /// let mut only_t_z = loc.clone();
+    /// only_t_z.extensions.retain_by_type(|t| {
+    ///     t == ExtensionType::Transform
+    ///         || t == ExtensionType::Other(b'z')
+    /// });
+    /// assert_eq!(only_t_z, "und-t-mul-z-zzz");
+    /// ```
+    pub fn retain_by_type<F>(&mut self, mut predicate: F)
+    where
+        F: FnMut(ExtensionType) -> bool,
```
Comment (zbraniecki) [inline components/locid/src/extensions/mod.rs:162]: `suggestion`: it seems that you don't need this to be mutable - you can use `Fn` here. What's more, you can just use `predicate: fn(usize) -> bool`.

Comment (sffc) [reply components/locid/src/extensions/mod.rs:162]: I based this signature on `Vec::retain`

https://doc.rust-lang.org/std/vec/struct.Vec.html#method.retain

> `fn(usize) -> bool`

what did you envision would be in the `usize`?

Comment (sffc) [reply components/locid/src/extensions/mod.rs:162]: CC @Manishearth

Comment (Manishearth) [reply components/locid/src/extensions/mod.rs:162]: In general for closures it's helpful to use FnMut where possible for maximum flexibility. Using FnMut is not constraining the user of this API, it constrains us (as implementors), wheras using Fn/FnOnce constrain the user while giving us more freedom.

Comment (zbraniecki) [reply components/locid/src/extensions/mod.rs:162]: gotcha. thanks!

**Review by zbraniecki — APPROVED** [review]

**Pull request merged**